src/v/cloud_topics/core/tests/write_pipeline_test.cc

```diff
@@ -0,0 +1,122 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "cloud_topics/core/write_pipeline.h"
+#include "model/fundamental.h"
+#include "model/namespace.h"
+#include "model/record_batch_reader.h"
+#include "test_utils/test.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/circular_buffer.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/manual_clock.hh>
+#include <seastar/core/sharded.hh>
+#include <seastar/core/when_all.hh>
+#include <seastar/util/later.hh>
+#include <seastar/util/log.hh>
+#include <seastar/util/noncopyable_function.hh>
+
+#include <chrono>
+
+namespace cloud_topics = experimental::cloud_topics;
+using namespace std::chrono_literals;
+
+namespace experimental::cloud_topics::core {
+struct write_pipeline_accessor {
+    // Returns true if the write request is in the `_pending` collection
+    bool write_requests_pending(size_t n) {
+        return pipeline->_pending.size() == n;
+    }
+
+    write_pipeline<ss::manual_clock>* pipeline;
+};
+} // namespace experimental::cloud_topics::core
+
+// Simulate sleep of certain duration and wait until the condition is met
+template<class Fn>
+ss::future<>
+sleep_until(std::chrono::milliseconds delta, Fn&& fn, int retry_limit = 100) {
+    ss::manual_clock::advance(delta);
+    for (int i = 0; i < retry_limit; i++) {
+        co_await ss::yield();
+        if (fn()) {
+            co_return;
+        }
+    }
+    GTEST_MESSAGE_("Test stalled", ::testing::TestPartResult::kFatalFailure);
+}
+
+TEST_CORO(write_pipeline_test, single_write_request) {
+    cloud_topics::core::write_pipeline<ss::manual_clock> pipeline;
+    cloud_topics::core::write_pipeline_accessor accessor{
+      .pipeline = &pipeline,
+    };
+    auto reader = model::make_empty_record_batch_reader();
+    // Expect single upload to be made
+
+    const auto timeout = 1s;
+    auto fut = pipeline.write_and_debounce(
+      model::controller_ntp, std::move(reader), timeout);
+
+    // Make sure the write request is in the _pending list
+    co_await sleep_until(
+      10ms, [&] { return accessor.write_requests_pending(1); });
+
+    auto res = pipeline.get_write_requests(
+      1, cloud_topics::core::initial_pipeline_stage);
+    ASSERT_TRUE_CORO(res.complete);
```

Comment:
I find it easier to ensure basic things are correct rather than be surprised later with bugs. I understand the argument and I find it acceptable if there is good degree of confidence that the alternative code paths are covered by some other tests.